main/adapters/repository/protocol.go

```diff
@@ -247,6 +283,72 @@ func (p *ExtendedProtocol) IsInitialized(uid uuid.UUID) (initialized bool, err e
 	return true, nil
 }
 
+func (p *ExtendedProtocol) VerifyBackendResponse(requestUPP, responseUPP []byte) (signatureOk bool, chainOk bool, err error) {
+	if !p.verifyNiomonResponse {
+		return false, false, nil
+	}
+
+	// check if backend response is a UPP or something else, like an error message string, for example "Timeout"
+	if !hasUPPHeaders(responseUPP) {
+		return signatureOk, chainOk, fmt.Errorf("response from UBIRCH Trust Service is not a UPP: %q", responseUPP)
+	}
+
+	signatureOk, err = p.verifyBackendResponseSignature(responseUPP)
+	if err != nil {
+		return signatureOk, chainOk, err
+	}
+
+	chainOk, err = p.verifyBackendResponseChain(requestUPP, responseUPP)
+	return signatureOk, chainOk, err
+}
+
+// hasUPPHeaders is a helper function to check if the data starts with the expected UPP headers
+func hasUPPHeaders(data []byte) bool {
+	return bytes.HasPrefix(data, signedUPPHeader) || bytes.HasPrefix(data, chainedUPPHeader)
+}
+
+// verifyBackendResponseSignature verifies the signature of the backend response UPP
+func (p *ExtendedProtocol) verifyBackendResponseSignature(upp []byte) (bool, error) {
+	if verified, err := p.Verify(p.backendUUID, upp); !verified {
+		if err != nil {
+			return false, fmt.Errorf("could not verify backend response signature: %v", err)
+		}
+		pub, _ := p.Crypto.GetPublicKeyBytes(p.backendUUID)
+		return false, fmt.Errorf("backend response signature verification failed with public key: %s",
+			base64.StdEncoding.EncodeToString(pub))
+	}
+	return true, nil
+}
+
+// verifyBackendResponseChain verifies that backend response previous signature matches signature of request UPP
+func (p *ExtendedProtocol) verifyBackendResponseChain(requestUPPBytes, responseUPPBytes []byte) (bool, error) {
+	requestUPP, err := ubirch.Decode(requestUPPBytes)
+	if err != nil {
+		// this shouldn't happen
```

Comment:
What do you mean that it shouldn't happen?

Reply:
as I am writing my answer to your question, I realize, that my thought process on this wasn't correct. It shouldn't happen because we are creating and encoding this UPP in this application, so being unable to decode it again at this point would indicate a bug in the application. The comment however is not useful here, so i will remove it.